### ISSUE TITLE
GridSearchCV error

### ISSUE THREAD

**Issue by vishalseshagiri** [fracpete/python-weka-wrapper#58]
```
from weka.classifiers import GridSearch
grid = GridSearch(options=["-sample-size", "100.0", "-traversal", "ROW-WISE", "-num-slots", "1", "-S", "1"])
grid.evaluation = "CC"
grid.y = {"property": "kernel.gamma", "min": -3.0, "max": 3.0, "step": 1.0, "base": 10.0, "expression": "pow(BASE,I)"}
grid.x = {"property": "C", "min": -3.0, "max": 3.0, "step": 1.0, "base": 10.0, "expression": "pow(BASE,I)"}
cls = Classifier(
    classname="weka.classifiers.functions.SMOreg",
    options=["-K", "weka.classifiers.functions.supportVector.RBFKernel"])
grid.classifier = cls
grid.build_classifier(train)
print("Model:\n" + str(grid))
print("\nBest setup:\n" + grid.best.to_commandline())
```
This code snippet from your documentation gives me this error
```
Exception: Failed to instantiate GridSearch - package installed and jvm started with package support?
```
All other classifiers are importing and working exactly as they should. Am I doing something wrong ?

**Comment by fracpete** [comment]
Did you start up the jvm with package support, like the error message suggested?
`jvm.start(packages=True)`

**Comment by fracpete** [comment]
Closed due to inactivity

**Comment by naiayti** [comment]
I have the same problem. I cannot understand what I am doing wrong as I am following strictly the example and I have jvm.start(packages=True) 
Could you please help?

**Comment by fracpete** [comment]
Did you install the `gridSearch` package? GridSearch is not part of core Weka.

**Comment by naiayti** [comment]
How can I install gridSearch? I have already the python-weka-wrapper3 and weka.classifiers working

**Comment by fracpete** [comment]
Either through the Weka GUI or with pww itself:
http://fracpete.github.io/python-weka-wrapper/api.html#packages

NB: You should switch to pww3, as Python 2.7 has reached its end of life:
http://fracpete.github.io/python-weka-wrapper3

**Comment by eilkou** [comment]
Thanks. I download it the package with 
`packages.install_package("gridsearch")` and I dont have the **Exception: Failed to instantiate GridSearch - package installed and jvm started with package support?** anymore.

Now I have an issue with the **iter**. I get
TypeError: iter() returned non-iterator of type 'JavaArrayIterator'

Also, my predictors have an issue with the iter now as I get errors:
TypeError: iter() returned non-iterator of type 'AttributeIterator'

and
TypeError: iter() returned non-iterator of type 'InstanceIterator'

They might be unrelated because they seem to be triggered on different parts of the code, but I really do not know how to fix the iter issues. Do you have any idea?

**Comment by fracpete** [comment]
Can you please post the code that generates this exception?

**Comment by fracpete** [comment]
BTW You should switch to [python-weka-wrapper3](https://github.com/fracpete/python-weka-wrapper3), the Python 3 version of this library. Python 2.7 (and this library) have reached their end-of-life.

**Comment by eilkou** [comment]
Thank you the problem was that I had installed both python-weka-wrapper and python-weka-wrapper3. I uninstalled both and reinstalled the python-weka-wrapper3. That fixed all the iter errors I had. 

One more question, how do I setup a gridsearch for weka.classifiers.trees.J48? 
I am trying to build it based on the previous example for support vector, but I get error for C property.

`grid = GridSearch(options=["-sample-size", "100.0", "-traversal", "COLUMN-WISE", "-num-slots", "1", "-S", "1"])
grid.evaluation = "RMSE"
grid.y = {"property": "-M", "min": 2, "max": 5, "step": 1.0, "expression": "I"}
grid.x = {"property": "-C", "min": 0.1, "max": 0.4, "step": 0.1, "expression": "I"}

cls = Classifier(
    classname= "weka.classifiers.trees.J48" , 
    options=None
    )
grid.classifier = cls
grid.build_classifier(data)
print("Model:\n" + str(grid))
print("\nBest setup:\n" + grid.best.to_commandline())`

It gives `JavaException: Thread-based execution of evaluation tasks failed: Failed to set property -C on object weka.classifiers.trees.J48`

P.S. It also prints **Failed to get class weka.classifiers.meta.GridSearch** but so does one I run it for the support vector, which compiles ok.

**Comment by fracpete** [comment]
You are supplying command-line option flags, not property names.

The properties that GridSearch/MultiSearch are to optimize are defined by a path made up of Java Bean property names. Unfortunately with Weka, there are two ways of configuring an object: command-line options and [Java Bean properties](https://docs.oracle.com/javase/tutorial/javabeans/writing/properties.html) (used by the Weka GUI). The command-line options cannot be used by GridSearch/MultiSearch as they are not typed and only represent a single level (eg you couldn't change the parameters of an SMO kernel). The Bean properties approach gets around the problem, but, unfortunately, uses different names that you usually determine through using the Weka GUI, like the Weka Explorer. 

I've just pushed out a new release with an added method that allows you to list all the property names from an object:
http://fracpete.github.io/python-weka-wrapper3/examples.html#parameter-optimization-property-names

Another source to figure out the property names manually, is by studying the [Javadoc](https://weka.sourceforge.io/doc.dev/) of the relevant class and its methods. A get/set pair like `getMinNumObj` and `setMinNumObj` results in a property name of `minNumObj`.

Long story short, your code should read:
```python
grid.y = {"property": "minNumObj", "min": 2, "max": 5, "step": 1.0, "expression": "I"}
grid.x = {"property": "confidenceFactor", "min": 0.1, "max": 0.4, "step": 0.1, "expression": "I"}
```

For further questions, please use the mailing list:
https://groups.google.com/forum/#!forum/python-weka-wrapper
